drivers/adc/iadc_gecko.c

```diff
@@ -0,0 +1,482 @@
+};
+
+struct adc_gecko_config {
+	IADC_Config_t config;
+	IADC_TypeDef *base;
+	void (*irq_cfg_func)(void);
+};
+
+static void adc_gecko_set_config(const struct device *dev)
+{
+	struct adc_gecko_data *data = dev->data;
+	struct adc_gecko_channel_config *channel_config = NULL;
+	const struct adc_gecko_config *config = dev->config;
+
+	IADC_TypeDef *iadc = (IADC_TypeDef *)config->base;
+	IADC_InitSingle_t sInit = IADC_INITSINGLE_DEFAULT;
+	IADC_SingleInput_t initSingleInput = IADC_SINGLEINPUT_DEFAULT;
+	IADC_Init_t init = IADC_INIT_DEFAULT;
+	IADC_AllConfigs_t initAllConfigs = IADC_ALLCONFIGS_DEFAULT;
+
+	channel_config = &data->channel_config[data->channel_id];
+
+	initSingleInput.posInput = channel_config->input_positive;
+	initSingleInput.negInput = channel_config->input_negative;
+
+	initAllConfigs.configs[0].analogGain = channel_config->gain;
+
+	initAllConfigs.configs[0].reference = channel_config->reference;
+
+	IADC_init(iadc, &init, &initAllConfigs);
+
+	IADC_initSingle(iadc, &sInit, &initSingleInput);
+}
+
+static int adc_gecko_check_buffer_size(const struct adc_sequence *sequence,
+					uint8_t active_channels)
+{
+	size_t needed_buffer_size;
+
+	needed_buffer_size = active_channels * sizeof(uint16_t);
+
+	if (sequence->options) {
+		needed_buffer_size *= (1 + sequence->options->extra_samplings);
+	}
+
+	if (sequence->buffer_size < needed_buffer_size) {
+		LOG_DBG("Provided buffer is too small (%u/%u)",
+			sequence->buffer_size, needed_buffer_size);
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+static int adc_gecko_check_resolution(const struct adc_sequence *sequence)
+{
+	int value = sequence->resolution;
+
+	/* Base resolution is on 12, it can be changed only up by oversampling */
+	if (value != 12) {
+		return -EINVAL;
+	}
+
+	return value;
+}
+
+static int start_read(const struct device *dev, const struct adc_sequence *sequence)
+{
+
+	struct adc_gecko_data *data = dev->data;
+	uint32_t channels;
+	uint8_t channel_count;
+	uint8_t index;
+	int res;
+
+	/* Check if at least 1 channel is requested */
+	if (sequence->channels == 0) {
+		LOG_DBG("No channel requested");
+		return -EINVAL;
+	}
+
+	if (sequence->oversampling) {
+		LOG_ERR("Oversampling is not supported");
+		return -ENOTSUP;
+	}
+
+	/* Check resolution setting */
+	res = adc_gecko_check_resolution(sequence);
+	if (res < 0) {
+		return -EINVAL;
+	}
+	data->resolution = res;
+
+	/* Verify all requested channels are initialized and store resolution */
+	channels = sequence->channels;
+	channel_count = 0;
+	while (channels) {
+		/* Iterate through all channels and check if they are initialized */
+		index = find_lsb_set(channels) - 1;
+		if (!data->channel_config[index].initialized) {
+			LOG_DBG("Channel not initialized");
+			return -EINVAL;
+		}
+		channel_count++;
+		channels &= ~BIT(index);
+	}
+
+	/* Check buffer size */
+	res = adc_gecko_check_buffer_size(sequence, channel_count);
+	if (res < 0) {
+		return res;
+	}
+
+	adc_gecko_set_config(data->dev);
```

Comment:
Why is it called here when `data->channel_id` has not been set yet?

Reply:
You're right that it shouldn't be called here. Removed.